trempy/shared/shared_auxiliary.py

```diff
@@ -19,45 +20,52 @@
-        df_subset['likl_lower'] = rv.cdf(lower)
-
-        df_subset['likl'] = 0.0
-        df_subset['likl'][is_upper] = df_subset['likl_upper'].loc[is_upper]
-        df_subset['likl'][is_lower] = df_subset['likl_lower'].loc[is_lower]
-        df_subset['likl'][is_not] = df_subset['likl_not'].loc[is_not]
-
-        contribs += df_subset['likl'].values.tolist()
-
-    rslt = -np.mean(np.log(np.clip(sorted(contribs), TINY_FLOAT, np.inf)))
-
-    return rslt
+    """Calculate the likelihood of the observed sample."""
+    m_optimal = get_optimal_compensations(
+        version=version, paras_obj=paras_obj, questions=questions, **version_specific)
+    data = copy.deepcopy(df)
+
+    # Add auxiliary data (question cutoffs, decision implied by the model, std of observed choices)
+    df_cutoff = pd.DataFrame.from_dict(cutoffs, orient='index', columns=['lower', 'upper'])
+    df_cutoff.index.name = 'Question'
+    data = data.join(df_cutoff, how='left')
+
+    df_m_optimal = pd.DataFrame.from_dict(m_optimal, orient='index', columns=['m_optim'])
+    df_m_optimal.index.name = 'Question'
+    data = data.join(df_m_optimal, how='left')
+
+    df_std = pd.DataFrame(sds, index=questions, columns=['std'])
+    df_std.index.name = 'Question'
+    data = data.join(df_std, how='left')
+
+    # Subjects who selected both Option A and B at least once. This implies their valuation
+    # is in the left-open interval (lower, upper], i.e. they initially prefered Option A at 'lower',
+    # but chose option B when it offered 'upper'.
+    data['is_interior'] = (data.lower < data.Compensation) & (data.Compensation < data.upper)
+    # Subjects who always prefered option A, i.e. their value of option A is higher than 'upper'.
+    data['is_upper'] = ((data['Compensation'].isin([NEVER_SWITCHERS])) |
+                        (data.Compensation > data.upper))
+    # Subjects who always prefered option B. So their value of Option A is smaller than 'lower'
+    data['is_lower'] = (data.Compensation <= data.lower)
+
+    # We only need the standard normal distribution for standardized choices.
+    data['choice_standardized'] = (data['Compensation'] - data['m_optim']) / data['std']
+    data['lower_standardized'] = (data['lower'] - data['m_optim']) / data['std']
+    data['upper_standardized'] = (data['upper'] - data['m_optim']) / data['std']
+    rv = norm(loc=0.0, scale=1.0)
+
+    # Likelihood: pdf for interior choices
+    likl_interior = (rv.pdf(data['choice_standardized'].loc[data['is_interior']]) /
+                     data['std'].loc[data['is_interior']])
+
+    # Likelihood: cdf for indifference points that are outside our choice list.
+    likl_upper = 1.0 - rv.cdf(data['upper_standardized'].loc[data['is_upper']])
+    likl_lower = rv.cdf(data['lower_standardized'].loc[data['is_lower']])
+
+    # Average negative log-likelihood
+    contribs = likl_interior.tolist() + likl_lower.tolist() + likl_upper.tolist()
```

Comment:
It would be nice to avoid this type changes, by using for example

```
contribs = np.concatenate([likl_interior.values, likl_lower, likl_upper], axis=0)
rslt = - np.mean(np.log(np.clip(np.sort(contribs), TINY_FLOAT, np.inf)))
```